### ISSUE TITLE
Allowing user to keep custom text if value not found from dropdown lists

### ISSUE THREAD

**Issue by jiteshsojitra** [harvesthq/chosen#2895]
Allowing user to keep custom text if value not found from dropdown lists.

My requirement is input value should preserve the text if value not found from dropdown lists.

### Steps to reproduce

1. Assume dropdown value contains:
India
USA

2. Type UK

3. Move mouse outside dropwown

Actual: UK is removed because dropdown doesn't find it.

Expected: Preserve UK text as it is to use it for other purpose

Any work around? JS customization help? - Highly appreciated.

**Comment by Skeeve** [comment]
The issue here is - as I can see it - If you have India and USA in the list of options, the actual values could be not "India" and "USA" but "91" and 1". So if you want to add "UK", how should the value "44" be known?

Just to clarify: I'm talking about the international call prefixes.

**Comment by jiteshsojitra** [comment]
I'm sorry but in my example i'm talking about select values present in dropdown list. UK text is my custom if some user want to specify it in the input field. Right now the problem is, when you put focus outside immediately Chosen field clears that value. I'm expecting Chosen field keeps value as it is as a input field as well to re-use as custom entered value.

**Comment by Skeeve** [comment]
Sorry, but it seems I'm unable to understand the use case. Maybe someone else can or maybe you can explaine differently?

**Comment by Mikk3lRo** [comment]
I understand the use case - it's like the tagging box in wordpress (among many others). Some call it "tagging", others "Dynamic Option Creation".

You are totally correct in saying that it makes no sense at all if your values are not related to the labels. 

But the new values are normally handled by having some "external" system to handle the creation and editing of tags - so in other words chosen would just insert a new option in the original select using the entered value for both the label and value and select that option. Then it would be up to the developer to handle this non-existent value on form submission, to actually add it to the "tag-list" or whereever the values are coming from...

Select2 has the feature:
https://select2.org/tagging

**Comment by jiteshsojitra** [comment]
I'm sorry, i'm saying something else; probably very easy to understand but we diverted communication to different feature. See this video - you will surely get it. Please re-read my issue to get more idea.

Repro steps video: https://www.screencast.com/t/gLFnxsnX

**Comment by Mikk3lRo** [comment]
Even after seeing your screencast it still seems to me like you are referring to "dynamic option creation"... Have you looked at the select2 link at the end of my previous post and tried that?

**Comment by jiteshsojitra** [comment]
Yes, Mikk; whatever link and information you provided completely makes sense. Thanks a lot!

I don't want to create the option, if it just preserves the value in dropdown list as input box then it also fine for me to use that custom value for other operation. At this point i like Chosen and don't want to switch to select2.

Need to play with java script a bit so it doesn't clear the value if it doesn't find from dropdown? Any help?

**Comment by Mikk3lRo** [comment]
I'm not suggesting you switch to select2 - it was simply meant as an example.

As far as I know there is currently no way to do it with (a "vanilla") chosen, though the feature has been requested several times. I may have a go at it if I have time, but I won't make any promises just yet.

Have you looked at the coffeescript files to see if you might be able to do it yourself? Or if someone has already made a fork that supports it?

**Comment by jiteshsojitra** [comment]
Agreed Mikk. The biggest problem with dynamic creation is:
- User may enter wrong and junk data and we can't save each and every values into DB (dropdown is picking up values from DB for professional software)
- Dropdown will flood with unwanted values
- Custom/floating input value is bother-less for admin if just used for other purpose and won't be saved

I'm sorry i didn't take a look coffeescript files but i tried without any luck, as i'm new to all these stuff. Appreciated Mikk for your attention and it would be great if you put some kind of temporary solution without clearing input value when focus goes off and long term solution you can improve it professionally like Select2.

**Comment by koenpunt** [comment]
#166 is an implementation of this.

**Comment by JpSilver** [comment]
> Agreed Mikk. The biggest problem with dynamic creation is:
> 
>     * User may enter wrong and junk data and we can't save each and every values into DB (dropdown is picking up values from DB for professional software)
> 
>     * Dropdown will flood with unwanted values
> 
>     * Custom/floating input value is bother-less for admin if just used for other purpose and won't be saved
> 
> 
> I'm sorry i didn't take a look coffeescript files but i tried without any luck, as i'm new to all these stuff. Appreciated Mikk for your attention and it would be great if you put some kind of temporary solution without clearing input value when focus goes off and long term solution you can improve it professionally like Select2.

Should this not be handled by the developer? I feel like this option would be a lot better than doing a Ajax call for the addition of the new value, lets say the user enters the wrong value, then you as the front and back end developer needs to add a field for removing this option where as if you enter a name eg Affected user, by the time you get to the end of the page you know that field is correct. If its not its the back-end developers work to check for this and display the error message. I feel like saying this is like saying hey we are to lazy to remove the auto clear feature. I mean we are developing websites with this so its OUR job to ensure the data that goes within that select is correct. I certainly don't want to add tags because that changes the whole database because i only want 1 option and not multiple. Is there really no way of just DISABLING the auto clear feature if no option is found?

**Comment by JpSilver** [comment]
> Agreed Mikk. The biggest problem with dynamic creation is:
> 
>     * User may enter wrong and junk data and we can't save each and every values into DB (dropdown is picking up values from DB for professional software)
> 
>     * Dropdown will flood with unwanted values
> 
>     * Custom/floating input value is bother-less for admin if just used for other purpose and won't be saved
> 
> 
> I'm sorry i didn't take a look coffeescript files but i tried without any luck, as i'm new to all these stuff. Appreciated Mikk for your attention and it would be great if you put some kind of temporary solution without clearing input value when focus goes off and long term solution you can improve it professionally like Select2.

I don't agree at all... Its the developers duty to check for these things. if you think like this you are making your modules less modular. You need to allow for freedom for the developer. At least allow for a disable option for the auto clear if no value is found. It doesn't even have to add to the list. If it can just stay in the text field then everything should be fine. The back-end developer can check for this and append the data to the database, when the page is reloaded the new value will appear in the select. I really don't get why you guys are saying these things... Where as if you "Hack" the select now and dynamically add the items to the options you will encounter these problems obviously but this will only be while the page is open or if you update the list via AJAX calls, exactly how you set it up now. So please re-conceder and just allow for "DON'T CLEAR THE FREAKEN LIST IF THERE ARE NO VALUES!"
